src/librustc_ast/ast.rs

```diff
@@ -314,35 +314,32 @@ pub type GenericBounds = Vec<GenericBound>;
 pub enum ParamKindOrd {
     Lifetime,
     Type,
-    // unordered is only true if sess.has_features().const_generics
-    // is active. Specifically, if it's only `min_const_generics`, it will still require ordered
-    // types.
+    // `unordered` is only `true` if `sess.has_features().const_generics`
+    // is active. Specifically, if it's only `min_const_generics`, it will still require
+    // ordering `const`s after `type`s.
     Const { unordered: bool },
 }
 
-impl PartialOrd for ParamKindOrd {
-    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+impl Ord for ParamKindOrd {
+    fn cmp(&self, other: &Self) -> Ordering {
         use ParamKindOrd::*;
-        let ord = match self {
-            &Lifetime if other == &Lifetime => Ordering::Equal,
-            &Lifetime => Ordering::Less,
-
-            &Type if other == &Lifetime => Ordering::Greater,
-            &Type if other == &Type || other == &Const { unordered: true } => Ordering::Equal,
-            &Type => Ordering::Less,
-
-            &Const { .. } if other == &Lifetime => Ordering::Greater,
-
-            &Const { unordered: false } if other == &Type => Ordering::Greater,
+        match (self, other) {
+            (&Lifetime, &Lifetime) => Ordering::Equal,
+            (&Lifetime, _) => Ordering::Less,
+            (_, &Lifetime) => Ordering::Greater,
+            (&Type | &Const { unordered: true }, &Type | &Const { unordered: true }) => {
+                Ordering::Equal
+            }
+            (&Type, &Const { unordered: false }) => Ordering::Less,
+            (&Const { unordered: false }, &Type) => Ordering::Greater,
 
-            &Const { .. } => Ordering::Equal,
-        };
-        Some(ord)
+            (&Const { .. }, &Const { .. }) => Ordering::Equal,
+        }
     }
 }
-impl Ord for ParamKindOrd {
-    fn cmp(&self, other: &Self) -> Ordering {
-        self.partial_cmp(other).unwrap()
+impl PartialOrd for ParamKindOrd {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(&other))
```

Comment:
```suggestion
        Some(self.cmp(other))
```
afaict the ref is unnecessary here